ccxt/async/exchange.py

```diff
@@ -72,6 +75,26 @@ def __del__(self):
     # def run_rest_poller_loop
     #     await asyncio.sleep (exchange.rateLimit / 1000.0)
 
+    async def wait_for_token(self):
+        while self.rateLimitTokens <= 1:
+            # if self.verbose:
+            #     print('Waiting for tokens: Exchange: {0}'.format(self.id))
+            self.add_new_tokens()
+            seconds_delays = [0.01, 0.1, 0.7, 1, 1.5, 2]
+            delay = random.choice(seconds_delays)
+            await asyncio.sleep(delay)
+        self.rateLimitTokens -= 1
+
+    def add_new_tokens(self):
+        # if self.verbose:
+        #     print('Adding new tokens: Exchange: {0}'.format(self.id))
+        now = time.monotonic()
+        time_since_update = now - self.rateLimitUpdateTime
+        new_tokens = math.floor( ( 0.8 * 1000.0 * time_since_update ) / self.rateLimit )
```

Comment:
... and this one, the `0.8` magic constant looks misterious ) I may be missing something here %)

Reply:
So initially new tokens was to set to be dynamically calculated, purposely based on how many tokens are initially sent out to adjust the token generation, so as to over a fixed time span (e.g. 60 seconds), the bucket algorithm to send the same number of requests a single delayed poller, sending out requests at a periodic time ( period of the rateLimit). Thus stay under the rate limit, and support bursts of requests over time, for optimal request rates etc.

The theory is as follows:

1. Traditional linear rate limit poller, rateLimit = 1.5 seconds, over a period for example T = 60 seconds, will send 60/1.5 = 40 requests, or here each request is represented as a token.
2. Dynamic rate limiter, must still satisfy sending 40 requests or less, over this same period of T = 60 seconds. Thus if the first burst, takes 16 requests, that leaves 40 - 16  = 24 requests left, of which should be added to the token bucket linearly, i.e. over this time period of T, thus a new token/request must be added every 2.5 seconds. This ensures that the original rate limit is met, whilst allowing bursts of requests over short periods of time. Thus the new_tokens = (1/2.5)* time_since_update.

These simple equations can be represented as:
<img width="880" alt="screen shot 2017-10-23 at 22 32 26" src="https://user-images.githubusercontent.com/6444377/31894850-22754f6c-b842-11e7-8e7e-3aca1d05e373.png">